src/components/checkbox/GroupableCheckboxReducers.ts

```diff
@@ -0,0 +1,143 @@
+  const nbSelected: number = state.nbSelected += (action.payload.checked ? 1 : 0);
+
+  if (state === groupableCheckboxInitialState) {
+    return {
+      ...state,
+      total,
+      nbSelected,
+      checkboxes: newChecboxes,
+    };
+  }
+
+  return _.extend(state, {
+    checkboxes: newChecboxes,
+    total,
+    nbSelected,
+  });
+};
+
+export const groupableCheckboxReducer = (state: IGroupableCheckboxesState = groupableCheckboxInitialState,
+                                         action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      if (action.payload.isParent) {
+        return addParentCheckbox(state, action);
+      }
+      return addChildCheckbox(state, action);
+
+    case GroupableCheckboxActions.toggleGroup:
+      if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+        let isChecked: boolean;
+        if (action.payload.isParent) {
+          isChecked = !state.parent.checked;
+          return {
+            ...state,
+            parent: {
+              id: action.payload.id,
+              checked: isChecked,
+            },
+            checkboxes: _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+              checkbox.checked = isChecked;
+              return checkbox;
+            }),
+            nbSelected: isChecked ? state.total : 0,
+          };
+        }
+        const checkboxes: ICheckboxState[] = _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+          if (checkbox.id === action.payload.id) {
+            isChecked = !checkbox.checked;
+            return {
+              id: checkbox.id,
+              checked: isChecked,
+            };
+          }
+          return checkbox;
+        });
+        if (checkboxes) {
+          const nbSelected: number = state.nbSelected += (isChecked ? 1 : -1);
+          return {
+            ...state,
+            parent: {
+              id: action.payload.parentId,
+              checked: nbSelected === state.total,
+            },
+            nbSelected,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
+      return state;
+    default:
+      return state;
+  }
+};
+
+export const groupableCheckboxesReducer = (state: IGroupableCheckboxesState[] = groupableCheckboxesInitialState,
+                                           action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState[] => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      const groupableCheckboxesState: IGroupableCheckboxesState = _.findWhere(state, (groupCheckboxState: IGroupableCheckboxesState) =>
+        groupCheckboxState.parentId === action.payload.id || groupCheckboxState.parentId === action.payload.parentId);
+      if (state && groupableCheckboxesState) {
+        groupableCheckboxReducer(groupableCheckboxesState, action);
+        return state;
```

Comment:
Shouldn't you return `groupableCheckboxReducer(groupableCheckboxesState, action);`?